src/v/cluster/persisted_stm.cc

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cluster/persisted_stm.h"
+
+#include "cluster/logger.h"
+#include "raft/errc.h"
+#include "raft/types.h"
+#include "storage/record_batch_builder.h"
+
+#include <seastar/core/future.hh>
+
+#include <filesystem>
+
+namespace cluster {
+
+persisted_stm::persisted_stm(
+  const char* snapshot_mgr_name,
+  ss::logger& logger,
+  raft::consensus* c,
+  config::configuration& config)
+  : raft::state_machine(c, logger, ss::default_priority_class())
+  , _c(c)
+  , _snapshot_mgr(
+      std::filesystem::path(c->log_config().work_directory()),
+      snapshot_mgr_name,
+      ss::default_priority_class())
+  , _log(logger)
+  , _config(config) {}
+
+ss::future<> persisted_stm::hydrate_snapshot(storage::snapshot_reader& reader) {
+    return reader.read_metadata().then([this, &reader](iobuf meta_buf) {
+        iobuf_parser meta_parser(std::move(meta_buf));
+        stm_snapshot_header hdr;
+        hdr.version = reflection::adl<int8_t>{}.from(meta_parser);
```

Comment:
It happens in the derivative classes inside load_snapshot. persisted_stm doesn't know about the actual format of the snapshots.

Reply:
if `persisted_stm::hydrate_snapshot` is reading the header why wouldn't it also verify the header? otherwise you'd duplicate that check in every sub-class?

Reply:
persisted_stm knows nothing about the structure of the snapshot from its perspective it's iobuf, version is for a derivative to let it know how to parse the data. It's like a snapshot in a snapshot.

Reply:
if you later added a crc to the header or some other flags how would future code be able to identify instances of old headers and act accordingly? it sounds like two versions are needed. at least some comment needs to be in the code about what exactly the version is describing. we are going to be in a few weeks starting to convert all of the existing serialization code over to a common framework and the general approach is: if it is a struct on disk it has a version, and that rule applies recursively.